frontend-react/e2e/pages/BasePage.ts

```diff
@@ -0,0 +1,202 @@
+    url: string,
+    fulfillOptions: RouteHandlerFulfillOptions,
+];
+
+export interface GotoRouteHandlerOptions {
+    mock?: RouteHandlers;
+    handlers?: RouteHandlers;
+    mockError?: boolean | RouteFulfillOptions;
+}
+
+export type BasePageTestArgs = TestArgs<"page" | "storageState">;
+
+export abstract class BasePage {
+    readonly page: Page;
+    readonly url: string;
+    readonly title: string;
+    readonly testArgs: BasePageTestArgs;
+    readonly routeHandlers: Map<string, Parameters<Page["route"]>[1]>;
+    protected _mockError?: RouteHandlerFulfillOptions;
+    protected _mockRouteCache: MockRouteCache;
+
+    readonly heading: Locator;
+
+    constructor(
+        { url, title, heading }: BasePageProps,
+        testArgs: BasePageTestArgs,
+    ) {
+        this.page = testArgs.page;
+        this.url = url;
+        this.title = title;
+        this.testArgs = testArgs;
+        this.routeHandlers = new Map();
+        this._mockRouteCache = {};
+        this.heading = heading ?? this.page.locator("INVALID");
+    }
+
+    get isMocked() {
+        return !this.testArgs.isMockDisabled;
+    }
+
+    get mockError(): RouteHandlerFulfillOptions | undefined {
+        return this._mockError;
+    }
+
+    set mockError(
+        err: boolean | number | RouteHandlerFulfillOptions | undefined,
+    ) {
+        if (err == null || err === false) {
+            this._mockError = undefined;
+            return;
+        }
+
+        this._mockError =
+            typeof err === "function" || typeof err === "object"
+                ? err
+                : {
+                      status: typeof err === "number" ? err : 500,
+                  };
+    }
+
+    get isErrorExpected() {
+        return !!this.mockError;
+    }
+
+    async reload() {
+        await this.route();
+
+        return await this.page.reload();
+    }
+
+    async goto(opts?: GotoOptions) {
+        await this.route();
+
+        return await this.page.goto(this.url, {
+            waitUntil: "domcontentloaded",
+            ...opts,
+        });
+    }
+
+    async route() {
+        for (const [url, handler] of this.routeHandlers.entries()) {
+            await this.page.route(url, handler);
+        }
+    }
+
+    addRouteHandlers(items: RouteHandlerEntry[]) {
+        for (const [url, _fulfillOptions] of items) {
+            const handler: RouteHandlerFn = async (route, request) => {
+                const { isMock, ...fulfillOptions } =
+                    typeof _fulfillOptions === "function"
+                        ? _fulfillOptions(request)
+                        : _fulfillOptions;
+                const mockErrorFulfillOptions = isMock
+                    ? typeof this.mockError === "function"
+                        ? this.mockError(request)
+                        : this.mockError
+                    : undefined;
+                const mockCacheFulfillOptions = isMock
+                    ? await this.getMockCacheFulfillOptions(url)
```

Comment:
when would it ever be undefined? since it's coming from a mock.

Reply:
there could be route handlers other than mocks (ex: the default ones) that don't even bother defining the property